# CPI sklearn model

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import accuracy_score
import numpy as np
from sklearn import metrics

In [8]:
df = pd.read_csv('../CPI_Data/Monthly_cpi_index.csv', index_col = 0)
df

Food  Food at home  Cereals and bakery products  \
2012-01-01  232.666       231.694                      266.677   
2012-02-01  232.486       231.180                      267.821   
2012-03-01  232.792       231.383                      267.101   
2012-04-01  233.234       231.711                      268.014   
2012-05-01  233.339       231.518                      268.653   
...             ...           ...                          ...   
2021-08-01  279.135       259.825                      288.990   
2021-09-01  281.506       262.695                      292.069   
2021-10-01  284.205       265.478                      294.625   
2021-11-01  285.507       266.384                      295.909   
2021-12-01  286.966       267.555                      297.279   

            Cereals and cereal products  Flour and prepared flour mixes  \
2012-01-01                      234.159                         252.159   
2012-02-01                      233.362                         250.564   
2012-03-01                      232.660                         252.104   
2012-04-01                      233.662                         252.102   
2012-05-01                      235.054                         254.336   
...                                 ...                             ...   
2021-08-01                      236.194                         248.823   
2021-09-01                      236.183                         250.198   
2021-10-01                      239.371                         249.956   
2021-11-01                      241.091                         246.663   
2021-12-01                      243.944                         248.322   

            Breakfast cereal  Rice, pasta, cornmeal  Bakery products    Bread  \
2012-01-01           228.664                242.992          283.880  170.658   
2012-02-01           227.984                242.217          286.484  172.978   
2012-03-01           227.997                238.975          285.771  172.955   
2012-04-01           228.862                240.693          286.589  173.559   
2012-05-01           230.326                241.615          286.629  173.581   
...                      ...                    ...              ...      ...   
2021-08-01           229.855                245.495          319.545  192.971   
2021-09-01           229.122                245.890          324.458  196.492   
2021-10-01           236.615                245.062          326.628  198.052   
2021-11-01           238.343                248.552          327.648  200.415   
2021-12-01           241.597                251.399          328.129  200.504   

            Fresh biscuits, rolls, muffins  ...  Transportation  \
2012-01-01                         167.911  ...         210.799   
2012-02-01                         168.364  ...         214.429   
2012-03-01                         168.406  ...         220.842   
2012-04-01                         166.211  ...         223.083   
2012-05-01                         167.158  ...         220.768   
...                                    ...  ...             ...   
2021-08-01                         186.280  ...         238.333   
2021-09-01                         194.852  ...         236.373   
2021-10-01                         195.203  ...         241.042   
2021-11-01                         197.481  ...         245.532   
2021-12-01                         194.413  ...         246.499   

            Private transportation  New and used motor vehicles  \
2012-01-01                 206.307                       99.659   
2012-02-01                 210.013                       99.889   
2012-03-01                 216.536                      100.325   
2012-04-01                 218.563                      100.977   
2012-05-01                 215.978                      101.399   
...                            ...                          ...   
2021-08-01                 237.994                      119.475   
2021-09

In [9]:
# nulls of ppi data
ppi = pd.read_csv('CPI_Data/PPI_data.csv', index_col = 0).T
for i in ppi.isna().sum():
    print(i)

FileNotFoundError: [Errno 2] No such file or directory: 'CPI_Data/PPI_data.csv'

In [10]:
#cant have null values in our model
df = df.dropna(axis = 1)
df

Food  Food at home  Cereals and bakery products  \
2012-01-01  232.666       231.694                      266.677   
2012-02-01  232.486       231.180                      267.821   
2012-03-01  232.792       231.383                      267.101   
2012-04-01  233.234       231.711                      268.014   
2012-05-01  233.339       231.518                      268.653   
...             ...           ...                          ...   
2021-08-01  279.135       259.825                      288.990   
2021-09-01  281.506       262.695                      292.069   
2021-10-01  284.205       265.478                      294.625   
2021-11-01  285.507       266.384                      295.909   
2021-12-01  286.966       267.555                      297.279   

            Cereals and cereal products  Flour and prepared flour mixes  \
2012-01-01                      234.159                         252.159   
2012-02-01                      233.362                         250.564   
2012-03-01                      232.660                         252.104   
2012-04-01                      233.662                         252.102   
2012-05-01                      235.054                         254.336   
...                                 ...                             ...   
2021-08-01                      236.194                         248.823   
2021-09-01                      236.183                         250.198   
2021-10-01                      239.371                         249.956   
2021-11-01                      241.091                         246.663   
2021-12-01                      243.944                         248.322   

            Breakfast cereal  Rice, pasta, cornmeal  Bakery products    Bread  \
2012-01-01           228.664                242.992          283.880  170.658   
2012-02-01           227.984                242.217          286.484  172.978   
2012-03-01           227.997                238.975          285.771  172.955   
2012-04-01           228.862                240.693          286.589  173.559   
2012-05-01           230.326                241.615          286.629  173.581   
...                      ...                    ...              ...      ...   
2021-08-01           229.855                245.495          319.545  192.971   
2021-09-01           229.122                245.890          324.458  196.492   
2021-10-01           236.615                245.062          326.628  198.052   
2021-11-01           238.343                248.552          327.648  200.415   
2021-12-01           241.597                251.399          328.129  200.504   

            Fresh biscuits, rolls, muffins  ...  Transportation  \
2012-01-01                         167.911  ...         210.799   
2012-02-01                         168.364  ...         214.429   
2012-03-01                         168.406  ...         220.842   
2012-04-01                         166.211  ...         223.083   
2012-05-01                         167.158  ...         220.768   
...                                    ...  ...             ...   
2021-08-01                         186.280  ...         238.333   
2021-09-01                         194.852  ...         236.373   
2021-10-01                         195.203  ...         241.042   
2021-11-01                         197.481  ...         245.532   
2021-12-01                         194.413  ...         246.499   

            Private transportation  New and used motor vehicles  \
2012-01-01                 206.307                       99.659   
2012-02-01                 210.013                       99.889   
2012-03-01                 216.536                      100.325   
2012-04-01                 218.563                      100.977   
2012-05-01                 215.978                      101.399   
...                            ...                          ...   
2021-08-01                 237.994                      119.475   
2021-09

In [11]:
#change index to datetime
df.index = pd.to_datetime(df.index)

In [12]:
# all the columns that we have in df
for i in df.columns:
    print(i)

Food
Food at home
Cereals and bakery products
Cereals and cereal products
Flour and prepared flour mixes
Breakfast cereal
Rice, pasta, cornmeal
Bakery products
Bread
Fresh biscuits, rolls, muffins
Cakes, cupcakes, and cookies
Other bakery products
Meats, poultry, fish, and eggs
Meats, poultry, and fish
Meats
Beef and veal
Uncooked ground beef
Uncooked beef roasts
Uncooked beef steaks
Uncooked other beef and veal
Pork
Bacon, breakfast sausage, and related products
Ham
Pork chops
Other pork including roasts and picnics
Other meats
Poultry
Chicken
Other poultry including turkey
Fish and seafood
Fresh fish and seafood
Processed fish and seafood
Eggs
Dairy and related products
Milk
Cheese and related products
Ice cream and related products
Other dairy and related products
Fruits and vegetables
Fresh fruits and vegetables
Fresh fruits
Apples
Bananas
Citrus fruits
Other fresh fruits
Fresh vegetables
Potatoes
Lettuce
Tomatoes
Other fresh vegetables
Processed fruits and vegetables
Canned fruits

In [13]:
#create x and y variables
x = df.drop(columns = ['All items less food','All items less shelter','All items less food and shelter',
                        'All items less food, shelter, and energy','All items less food, shelter, energy, and used cars and trucks',
                       'All items less medical care','All items less energy'], axis = 1)
y = df['All items less food']

In [14]:
#just the essential expenditure categories
x = df[['Food and beverages','Housing','Apparel','Transportation',
        'Medical care','Recreation','Education and communication',
        'Other goods and services']]

y = df['All items less food']

In [15]:
#split data into train and test
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.2, random_state = 0)

In [16]:
# linear regression object
lr = LinearRegression()
lr.fit(x_train,y_train)

LinearRegression()

In [17]:
y_predict = lr.predict(x_test)

In [18]:
y_predict

array([235.00438755, 256.89155133, 256.52292372, 231.01426559,
       258.27331646, 232.43579964, 229.80994069, 229.5494729 ,
       236.10655018, 255.86471879, 236.34261441, 236.56127749,
       228.8900462 , 241.87369241, 277.4694971 , 249.03595542,
       237.60657797, 236.88453279, 266.05274287, 251.21504552,
       243.66137566, 240.41759397, 232.43887622, 233.19452409])

In [19]:
r2_score = lr.score(x_test,y_test)
r2_score

0.9999744404112334

This is very odd, our model should not have such a high accuracy score.

This may be because our model is overfitting.

OR because there is a feature that is leaking information, giving the answer away.

Reasons why it may ne overfitting..
- using features that correlate greatly
- giving model too many features

What else can we do for extremely high accuracy scores?

look into why values are what they are and how theyre changing
-external factors

In [20]:
# make dataframe just the major expenditure categories
df = df[['Food and beverages','Housing','Apparel','Transportation',
        'Medical care','Recreation','Education and communication',
        'Other goods and services']]

In [21]:
# create a function that checks correlation for all columns to get the best model columns
def evaluate_model(model, x, y):
    scores = []
    x = x.dropna()
    y = y.dropna()
    for x_column in x:
        y = y.drop(columns = [x_column], axis = 1)
        for y_column in y:
            x_train, x_test, y_train, y_test = train_test_split(x[[x_column]], y[y_column], random_state=1, test_size=0.2)
            model.fit(x_train, y_train)
            predict = model.predict(x_test)
            score = np.sqrt(metrics.mean_squared_error(predict, y_test))
            scores.append(score)
            index_name = x_column + ' ' + y_column
            print(f"The model {model} can predict {index_name} columns with root mean squared error of {score}")

In [22]:
Linear_Reg = LinearRegression()
evaluate_model(Linear_Reg, df, df)

The model LinearRegression() can predict Food and beverages Housing columns with root mean squared error of 4.802734802268778
The model LinearRegression() can predict Food and beverages Apparel columns with root mean squared error of 3.109006378793539
The model LinearRegression() can predict Food and beverages Transportation columns with root mean squared error of 13.020549468619
The model LinearRegression() can predict Food and beverages Medical care columns with root mean squared error of 11.948399306573288
The model LinearRegression() can predict Food and beverages Recreation columns with root mean squared error of 0.8592326847729517
The model LinearRegression() can predict Food and beverages Education and communication columns with root mean squared error of 1.2894833481568584
The model LinearRegression() can predict Food and beverages Other goods and services columns with root mean squared error of 5.742651230742322
The model LinearRegression() can predict Housing Apparel columns 

Food and Beverage column with Recreation has the lowest RMSE score out of all

In [23]:
# run an accuracy score for the food and beverage column and recreation column
model = LinearRegression()
x_train, x_test, y_train, y_test = train_test_split(df[['Food and beverages']], df['Recreation'], random_state=1, test_size=0.2)
model.fit(x_train, y_train)
results = model.predict(x_test)
results

array([120.436951  , 117.4070011 , 117.37379549, 126.3140951 ,
       118.84383051, 117.55754986, 116.46575925, 119.13868638,
       117.65017605, 121.16522753, 120.95725552, 122.11820369,
       113.77210982, 120.24795363, 124.42586907, 117.50137495,
       122.71790208, 122.67895415, 118.14226828, 117.83143377,
       123.94176616, 114.77526891, 119.34341272, 123.14757774])

In [24]:
r2_score = model.score(x_test,y_test)
r2_score

0.9346923861630986

In [26]:
from sklearn.metrics import mean_squared_error
mean_squared_error(results, y_test)

0.7382808065821345

We get an accuracy of 93% for our linear regression model using food and beverage with recreation categories.